pkg/agent/clientset.go

```diff
@@ -211,23 +217,40 @@ func (cs *ClientSet) sync() {
 	}
 }
 
-func (cs *ClientSet) connectOnce() error {
-	if !cs.syncForever && cs.serverCount != 0 && cs.ClientsCount() >= cs.serverCount {
-		return nil
+func (cs *ClientSet) ServerCount(ctx context.Context) int {
+	var serverCount int
+	if cs.leaseCounter != nil {
+		serverCount = cs.leaseCounter.Count(ctx)
+		if serverCount == 0 {
+			klog.Warningf("server lease counter could not find any leases")
+		}
+	} else {
+		serverCount = cs.lastReceivedServerCount
 	}
-	c, serverCount, err := cs.newAgentClient()
-	if err != nil {
-		return err
+
+	if serverCount != cs.lastServerCount {
+		klog.Warningf("change detected in proxy server count (was: %d, now: %d)", cs.lastServerCount, serverCount)
+		cs.lastServerCount = serverCount
 	}
-	if cs.serverCount != 0 && cs.serverCount != serverCount {
-		klog.V(2).InfoS("Server count change suggestion by server",
-			"current", cs.serverCount, "serverID", c.serverID, "actual", serverCount)
 
+	metrics.Metrics.SetServerCount(serverCount)
```

Comment:
Ideally we should record the metrics where get obtain them. The primary reason I say this is that for there are no leases case this will record a metric of 1 not 0.

Reply:
Hmm, don't we want the metric to report what the agent *thinks* the server count is, though? The agent will never believe the count is 0.

Happy to add a separate metric for the lease counting system!